File: Day4/find_the_kth_character_in_string_game_ii.js
function kthCharacter(k, operations) {
    let prevString = "a";
    let i = 0;

    while (prevString.length < k) {
        let currString = "";
        if (operations[i]) {
            for (let ch of prevString) {
                let nextChar = String.fromCharCode(
                    'a'.charCodeAt(0) + ((ch.charCodeAt(0) - 'a'.charCodeAt(0) + 1) % 26)
                );
                currString += nextChar;
            }
        } else {
            currString = prevString;
        }
        prevString += currString;
        i++;
    }

    return prevString[k - 1];
}

File: July/Day4/find_the_kth_character_in_string_game_ii.py
class Solution:
    #Tc - 2^t , t is the power 1, 2, 3,
    def kthCharacter(self, k: int) -> str:
        prevString, i = 'a', 0
        while len(prevString) < k:
            size = len(prevString)
            currString =''
            if operations[i]:
                for ch in prevString:
                    nextChar = chr(ord('a') + ((ord(ch) - ord('a') + 1) % 26))
                    currString+=nextChar
            else:
                currString = prevString
            prevString+= currString
            i+=1
        return prevString[k-1] 

class Solution:
    def kthCharacter(self, k: int, operations: List[int]) -> str:
        #TC = log(k)
        ans = 0
        while k != 1:
            power = k.bit_length() - 1
            if (1 << power) == k:
                power -= 1
            k -= 1 << power
            if operations[power]:
                ans += 1
        return chr(ord("a") + ans)

class Solution:
    #Tc = O(logK)
    def kthCharacter(self, k: int, operations: List[int]) -> str:
        ans = 0
        k-=1
        for bitPos in range(k.bit_length() - 1, -1, -1):
            if (k >> bitPos) & 1:
                ans += operations[bitPos]
        return chr(ord("a") + (ans % 26))
        
